### PR TITLE
Containerisation of Continuous Integration

### PR DESCRIPTION
**Pull request by tomtrafford**
These changes create a Container with all the tools required for developing any of the PandABlocks repos. This is uploaded as a package to Github and can be pulled from ghcr.io. The PandABlocks-rootfs CI has been updated to run in this container.

### CODE REVIEW

File: .github/workflows/code.yml [hunk 1]
```diff
@@ -83,3 +128,25 @@ jobs:
...
+        with:
+          name: pandablocks-dev-container
+          path: .
+    
+      - name: Load and push image to ghcr
+        run: |
+          docker load --input pandablocks-dev-container.tar
+          echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u PandABlocks --password-stdin
+          docker tag pandablocks-dev-container ghcr.io/pandablocks/pandablocks-dev-container:latest
+          docker push ghcr.io/pandablocks/pandablocks-dev-container:latest
```
Comment (coretl) [inline .github/workflows/code.yml:149]: We need to support tagging with versions too, have a chat to @gilesknap when he's back

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -42,7 +42,7 @@ jobs:
       # Publish docs
       - name: Move to versioned directory
         # e.g. master or 0.1.2
-        run: mv build/build/html "PandABlocks-rootfs/.github/pages/${{ steps.split.outputs._2 }}"
+        run: mv build/PandA-rootfs-zynq/build/html "PandABlocks-rootfs/.github/pages/${{ steps.split.outputs._2 }}"
```
Comment (coretl) [inline .github/workflows/docs.yml:45]: Can replace line 36-45 with something like https://github.com/epics-containers/python3-pip-skeleton/blob/10d0f042ce7cf83264cb35ae8622dd204d6a9df3/.github/workflows/docs.yml#L50-L52

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+# For the documentation
+RUN pip3 install matplotlib \ 
+    rst2pdf \
+    sphinx \
+    sphinx-rtd-theme \
+    --upgrade docutils==0.16
+
+# Create config file for dls-rootfs
+RUN bash scripts/config-file-rootfs.sh
```
Comment (coretl) [inline Dockerfile:55]: Isn't this `ln -s CONFIG.example CONFIG`?

**Review by coretl — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (tomtrafford) [reply Dockerfile:55]: This is creating the CONFIG.local file for the checked-out [rootfs ](https://github.com/dls-controls/rootfs) repo which it builds into the container. As that repo is not just used for PandA I thought we may not want to update its CONFIG.example with the PandA Container paths so I kept its creation from this script. If it's a better solution I can use its CONFIG.example and then update the paths?

Comment (coretl) [reply Dockerfile:55]: Ahh, I forgot we needed a CONFIG for rootfs too. This is fine then

File: .github/workflows/code.yml (hunk 1 above)
Comment (gilesknap) [reply .github/workflows/code.yml:149]: @tomtrafford this is my approach to building and pushing tagged container images (this example is for a Python project). 
https://github.com/epics-containers/python3-pip-skeleton/blob/main/.github/workflows/code.yml#L161-L217

Feel free to drop by to discuss.

**Comment by coretl** [comment]
We should mark releases as pre-release too:
https://github.com/epics-containers/python3-pip-skeleton/issues/34

**Comment by tomtrafford** [comment]
Hi, @gilesknap would you able to have a look through the CI for these changes, and let me know if you have any comments or if I should do anything differently, before I merge this in?

File: .github/workflows/code.yml [hunk 4]
```diff
@@ -79,7 +135,85 @@ jobs:
...
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Rebuild and release image
+        uses: docker/build-push-action@v2
+        with:
+          context: . 
+          file: ./PandABlocks-rootfs/Dockerfile
```
Comment (gilesknap) [inline .github/workflows/code.yml:215]: Have you worked out if it is quick to do the rebuild here?

If not maybe we could look at a different approach.

We really just need to do a docker push of the tarred artefact - but also would like to take advantage of Docker meta

File: .github/workflows/docs.yml [hunk 5]
```diff
@@ -7,6 +7,15 @@ on:
 jobs:
   build:
     runs-on: ubuntu-latest
+    # Conatiner from ghcr
+    container:
```
Comment (gilesknap) [inline .github/workflows/docs.yml:11]: Not seen this used before in GHA. nice!

**Review by gilesknap — APPROVED** [review]
All looks good to me.

File: .github/workflows/code.yml (hunk 4 above)
Comment (tomtrafford) [reply .github/workflows/code.yml:215]: Last time I ran the CI, the rebuild and release took 4m 30s whereas the original build took 7m34. When we downloaded the artifact it took 2m53 for one job and 6 m for the other... The time before that, the rebuild and release took 3m 34s, the original build took 7m 21, and the artifact downloads took 4m 14s and 4m 35s. I would say it definitely seems faster than redoing the original build again. It's hard to tell if its quicker than downloading the artifact again....

Comment (gilesknap) [reply .github/workflows/code.yml:215]: OK well then go with what you have.

**Pull request merged**